README.rst
============
SaberMod 2.0
============

This is a *Star Wars Jedi Knight II: Jedi Outcast* 1.04 mod targeting
all competitive communities. It is developed by *SaberMod team* (see
Authors_) and draws inspiration from all great id tech 3 games and
mods, trying to improve on usability, stability and user experience
with *no-gimmicks* approach. Main goal of the mod is to promote and
improve competitive aspect of the game.

Source code is hosted on GitHub_ and based on `JK2 SDK GPL`_ - an
updated JK2 1.04 SDK.

What's New
==========

This is just a list of new features that require explanation. Refer to
CHANGELOG.rst and Git commit history for full log.

Client-Side
-----------

Console Commands
................

follow [first|second]
  Follow first or second best player on the server. Can be used to
  stay on spectator team in Tournament gametype.

motd
  Show server Message of the Day.

nextspecmode/prevspecmode
  Cycle through spectator modes when following another
  player. Currently available modes are:
  1. Follow another player from his POV. (default)
  2. You can freely control camera angles and range.

players [?|team]
  List players connected to server with some additional
  info. Optionally limit to members of *team*.

timein
  Resume paused game.

timeout
  Pause game for 60 seconds.

ragequit
  Reserved for God, do not use.

ready
  Toggle ready status during warmup.

seek [+|-][mm:]ss
  During demo playback seeks to *mm:ss* server time. With *+* or *-*
  prefix seeks *mm* minutes and *ss* seconds forwards or backwards.

Call Vote
.........

abort
  Abort the match and go back to warmup.

matchmode <0|1>
  Enable/Disable match mode - restrict spectator chat and require
  matching clientside. In round-based gametypes dead players may only
  follow their teammates.

mode [mode]
  Switch to one of admin-defined game modes.

nk
  No Kick. Disables kicking completely and turns on friendly fire.

wk
  With Kick. Kicking mechanics depends on *g_kickMethod* value. Also
  turns off friendly fire.

poll <question>
  Ask an arbitrary question to players.

remove <player>
  Remove *player* to spectator team.

shuffle
  Randomize teams.

teamsize <size>
  Set maximum team size to *size*. 0 means unlimited. No players will
  be removed.

Call Team Vote
..............

forfeit
  Forfeit the game.

Referee Commands
................

These commands can be used only by a registered referee or server
console (rcon).

abort
  Abort the match and go back to warmup.

allready
  Make all players ready in warmup or during intermission.

announce <message|motd>
  Print *message* or ingame message of the day on everyone's screen.

forceteam <player|all> <team>
  Move players between teams.

help
  List referee commands

(un)lockteam <teams>
  Prevent players from joining *teams*.

(un)pause [seconds]
  Pause match for a specified number of seconds or until unpaused.

referee <player>
  Make player a referee without losing own referee status.

unreferee <player|all>
  Remove referee status.

CGame Cvars
...........

handicap <x>
  Lower your max health to x and damage to x%.

cg_autoSave <bitmask>
  Automatically save game record.

  | 1: Save screenshot of the final scoreboard
  | 2: Save demo of the match

cg_autoSaveFormat <format>
  Format is a template for auto demo/screenshot file name, where
  following tokens will be replaced with current values:

  | [date] - Date in yyyy-mm-dd format
  | [time] - Time in hh-mm format
  | [gametype] - Server gametype
  | [name] - Player name
  | [server] - Server name
  | [map] - Map name

cg_chatBeep <0|1>
  Turn on/off chat beep.

cg_crosshairColor <hex>
  Force crosshair color using hexadecimal rgb color code. Fourth
  position can be used to set transparency. Eg ``#00ff00ff`` is green.

cg_crosshairIndicators <bitmask>
  Draw crosshair indicators. Sum values from the following list:

  =====================  =====================
  1 - Movement arrows    2 - Speedometer
  =====================  =====================

cg_crosshairIndicatorsSpec <bitmask>
  Same as above but only when following another player.

cg_damagePlums <0|1>
  When you hit an enemy, draw a small damage plum coming out of his
  torso. Works only if server has *g_damagePlums* enabled.

cg_darkenDeadBodies <0|1>
  Darken dead bodies outside of duel too.

cg_drawClock <0|1>
  Draw clock showing your local time.

cg_drawFollow <0|1>
  Draw large "Following <playername>" message.

cg_drawRewards <0|1>
  Draw rewards for outstanding moves. Requires ent's "Jedi Knight
  Rewards 2" assets.

cg_drawTimer <0|1|2>
  Draw game timer. 1 - count up, 2 - count down.

cg_duelGlow <0|1>
  Turn on/off duel glow.

cg_fastSeek <0|1>
  Use experimental fast seeking method (see *seek* console command).

cg_fixServerTime <0|1>
  Fix various engine issues on servers running for a few days.

cg_followKiller <0|1>
  When player you are following dies, switch to his killer.

cg_followPowerup <0|1>
  Automatically follow flag and powerup carriers.

cg_fovAspectAdjust <0|1>
  Change Field Of View calculations so that they don't disadvantage
  widescreen monitors. Instead of cropping top and bottom parts of the
  screen it's extended to the sides, compared to 4:3 display. Works
  only when *cg_widescreen* is enabled.

cg_drawSpectatorHints <0|1>
  Draw extra hints on new spectator features.

cg_privateDuel <0|1>
  Hide all other players and entities when duelling. Available only
  on server running JK2MV 1.2 or newer.

cg_smoothCamera <0|1>
  Fix camera warping while maintaining original feel in following
  scenarios: unstable fps, unstable connection, overloaded server,
  local server, high velocity movement, demo playback.

cg_smoothCameraFPS <fps>
  Emulate specific fps with *cg_smoothCamera*. When this is 0, current
  com_maxfps is used instead. Useful for demo rendering.

cg_widescreen <0|1>
  Enable HUD adjustments for widescreen monitors

UI Cvars
........

ui_widescreen <0|1>
  Enable menu adjustments for widescreen monitors

Spectating
..........

As a spectator, *+use* button makes you change followed player using
"smart cycle" mode. It will switch between duelling players, search
for a next powerup player or cycle through current team in a
scoreboard order.

As a free floating spectator you can target a player with your
crosshair and press *+attack* button to start following him.

Server-Side
-----------

Console Commands
................

All `Referee Commands`_ can be used as console commands.

items [items]
  Enable/Disable items using human readable names. Type without
  argument to see usage instructions.

mode <mode|default>
  Change to *mode* or list all available modes when passed without
  arguments.

players [team]
  Print various informations about players. Optionally filter by team.

referee <password>
  Become a referee using password provided by server admin.

remove <player|all> [time]
  Remove *player* to spectator team for at least *time* seconds.

spawnitems [items]
  Enable/Disable spawning items using human readable names. Type
  without argument to see usage instructions.

shuffle
  Randomize teams.

Game Cvars
..........

teamsize <size>
  See `Call Vote`_ teamsize.

roundlimit <limit>
  Number of rounds in a round-based match.

lifelimit <limit>
  Number of lives per round in a round-based match.

duel_fraglimit
  Removed. Use roundlimit instead.

dmflags <bitmask>
  Sum of values from the following list:

  =====================  =====================  =====================
  1 - Fix jump height    2 - CJK Disruptor      4 - Go through team
  8 - No fall damage     16 - Limit FOV (97)    32 - No footsteps
  64 - No kick mode      128 - league mod YDFA
  =====================  =====================  =====================

bot_nochat <0|1>
  Prevent bots from sending chat messages.

g_allowRefVote <0|1|bitmask>
  Control what commands are available to referees. Uses the same
  bitmask as g_allowVote below.

g_allowTeamVote <bitmask>
  What team votes should be allowed. Team votes are called with
  *callteamvote* command.

  =====================  =====================  =====================
  2 - Team Leader        4 - Forfeit Match
  =====================  =====================  =====================

g_allowVote <0|1|bitmask>
  0 / 1 - disable / enable all votes.

  Moreover you can decide what votes should be available by setting
  it to a sum of values from the following list:

  =====================  =====================  =====================
  2 - Map Restart        4 - Next Map           8 - Map
  16 - Gametype          32 - Kick              64 - Shuffle
  128 - Do Warmup        256 - Timelimit        512 - Fraglimit
  1024 - Roundlimit      2048 - Teamsize        4096 - Remove
  8192 - WK/NK           16384 - Mode           32768 - Match Mode
  65536 - Capturelimit   131072 - Poll          262144 - Referee
  524288 - Abort         524288 - Lifelimit
  =====================  =====================  =====================

g_antiWarp <0|1|2>
  Prevention system against players who are warping or using lag scripts.

  | 1: Draw icon above warping player's head.
  | 2: Forcefully prevent players from warping for others. This
       setting makes game almost unplayable for a warping player and
       may hurt legitimate players who have bad connection.

  Refer to *g_antiWarpTime* cvar description for more details.

g_antiWarpTime <msec>
  Tune when player is considered as warping and g_antiWarp preventive
  actions are taken against him. Default setting is 1000 and it only
  marks players with interrupted connection. To prevent warping and
  lag scripts it should be set as low as possible so that legitimate
  players are not affected.

g_damagePlums <0|1>
  Allow clients with *cg_damagePlums* enabled to see damage plums.

g_dismember <percentage>
  Chance to dismemeber player killed with a lightsaber.

g_infiniteAmmo <0|1>
  Players spawn with infinite ammo for all weapons.

g_ingameMotd <message|none>
  Ingame message of the day shown to all players. May contain ``\n``
  for newline and ``\\`` for backslash.

g_instagib <0|1>
  Enable simple instagib mode for all weapons. Splash does no damage.

g_kickMethod <method>
  Choose one of following force kick methods:

  =====================  =====================  =====================
  0 - No effect          1 - Basejk             2 - No damage
  3 - League Mod
  =====================  =====================  =====================

g_log[1-4] <filename>
  You can use 4 separate log files now.

g_consoleFilter <mask>

g_logFilter[1-4] <mask>
  Filter events that should be printed in the dedicated server console
  or saved in the corresponding log file using following bit mask:

  =====================  =====================  =====================
  1 - Game Status        2 - Client Connect     4 - Client Begin
  8 - Userinfo Change    16 - Client Rename     32 - Client Spawn
  64 - Private Duel      128 - Obituary         256 - Say
  512 - Say Team         1024 - Tell            2048 - Voice Tell
  4096 - Item Pickup     8192 - Flag            16384 - Weapon Stats
  32768 - Game Stats     65536 - Duel Stats     131072 - Vote
  262144 - Referee Cmds
  =====================  =====================  =====================

g_macroscan <0|1>
  Enable scanning for and disabling binds that may give unfair
  advantage. Works only on players using SaberMod Clientside.

g_maxGameClients <limit>
  Removed. Use teamsize instead.

g_modeDefault <mode>
  Default server mode. Read `Server Modes`_ section to learn how to
  use it properly.

g_modeDefaultMap <map>
  Map for default mode. Leave blank to not change map.

g_modeIdleTime <minutes>
  Reset to default mode if server has been idle for this many minutes.

g_pushableItems <mask>
  What types of items should be movable with force push and pull:

  =====================  =====================  =====================
  2 - Weapon             4 - Ammo               8 - Armor
  16 - Health            32 - Powerup           64 - Holdable
  =====================  =====================  =====================

g_refereePassword <password>
  Allow players who know password to become referees using *referee*
  console command. When this cvar is empty (default), *referee*
  console command cannot be used to become a referee.

g_removeUnreachableItems <0|1>
  Remove items, bodies and flags that land in map areas inaccessible
  to players. This is a workaround for retail maps not doing a very
  good job at it. On a correctly made map it should not be needed.

g_requireClientside <0|1>
  Allow only players with matching clientside to join the game.

g_restrictChat <0|1>
  Prevent spectators from speaking to players and all clients from
  speaking to dueling players.

g_restrictSpectator <0|1>
  Dead players may only follow their teammates.

g_roundWarmup <seconds>
  How many seconds players get to reposition themselves at the start
  of a round.

g_spawnItems <bitmask>
  What items will be given to players on spawn. Use following bitmask:

  =====================  =====================  =====================
  2 - Seeker Drone       4 - Forcefield         8 - Bacta
  32 - Binoculars        64 - Sentry
  =====================  =====================  =====================

g_spawnShield <ammount>
  Ammount of shield player gets on spawn.

g_spawnWeapons <bitmask>
  Controls weapons given to players on spawn using the same bitmask
  as *g_weaponDisable*. The later cvar affects only weapons and ammo
  spawned on a map. Setting this cvar to 0 restores original behaviour
  of *g_weaponDisable*.

g_teamForceBalance <number>
  Prevents players from joining the weaker team if difference
  is greater than *number*.

g_teamsizeMin <size>
  Minimum votable teamsize.

g_timeoutDuration <seconds>
  Duration of a player timeout.

g_timeoutLimit <number>
  Maximum number of times a player is allowed to call a timeout.

g_unlagged <0|1|2>
  Experimental "unlagged" disruptor hit detection. 2 accounts for
  doors and other movers too at some server performance penalty.

g_unlaggedMaxPing <msec>
  Maximum lag compensation. Unlagged has subjective, counter-intuitive
  side effects. For example a player can be hit some time after he hid
  behind an obstacle. This cvar's value limits time period in which
  this can happen, adding extra hit detection delay for players with
  pings higher than *msec*.

g_warmup <0|1>
  SaberMod has a new warmup system. All players must ready up with
  *ready* command before a match can start. Old *g_warmupTime* Cvar is
  no longer used. Setting this cvar to 0 disables warmup alltogether.

g_vampiricDamage <ratio>
  Players become vampires and they drain health when damaging
  enemy. *ratio* value determines how much health is drained for 1
  point of damage. It may be set to a negative value.

g_voteCooldown <seconds>
  How long a player has to wait before he can call another vote.

Round-Based Gametypes
.....................

In round-based gametypes players spawn with all available weapons and
items (controlled by *g_spawnWeapons* and *g_spawnItems* cvars),
however there are no pickups on the map. Players gain one point for
killing an enemy and one point for each 50 damage dealt to the enemy
team. A round lasts until either one team is eliminated or a timelimit
is hit. Match ends when a roundlimit is hit. Lifelimit is a new
setting that allows players to respawn in a single run until their
lives run out.

Red Rover (g_gametype 9)
  It can be described as FFA with a twist. There are two teams, player
  who gets killed respawns in the opposing team. Round ends when one
  team is eliminated, but the match winner is a person who scores most
  points.

Clan Arena (g_gametype 10)
  Player who dies must spectate until the end of a round. When one
  team is eliminated, round is over. Team who hits the round limit
  first wins the match.

Last Man Standing (g_gametype 11)
  Player who dies must spectate until the end of a round. When
  everyone else is dead, last man standing receives 1 point and the
  game progresses to next round. There are no points for killing
  enemies or dealing damage.

Server Modes
............

Server administrator can configure a number of custom game "modes",
players will be able to choose from. A mode is technically a config
file in ``modes/`` directory that will be executed when players
sucessfuly vote to use it. It can contain any commands altering server
behaviour, but please take following guides into consideration.

Switching to a mode from any other should always result in the same
server state. To achieve this it's best to use a "reset" config,
executed at the start of each mode config. It should contain a default
value for every possible cvar your modes are changing. Examine
included modes and ``reset.cfg`` as an example.

Server can be configured to go back to a default mode after a period
of inactivity. To do so last lines of the main server config should
resemble following template::

  set g_modeIdleTime "10"
  set g_modeDefault "mymode"
  exec "modes/mymode"
  map ffa_bespin

Where *mymode* is the default mode.

Building SaberMod
=================

SaberMod has several build scripts, but only GNU Make script is
complete and used for official releases. other scripts produce
different QVM files from the same source code.

GNU Make
--------

Only requirements are GNU Make, Git and GCC or Clang
compiler. Makefile has several targets, type ``make <target>`` to
build selected target. Most useful are:

- **shared** build shared module libraries that should be used for
  debugging and development.
- **assets** create assets.pk3 - an amalgamation of all SaberMod
  clientside and serverside assets which is useful for development, as
  they will not be loaded from outside of a pk3.
- **serverside** create a full release zip package.

Type ``make help`` to learn about remaining targets.

Microsoft Visual Studio
-----------------------

Install Git command line program on your system. It can be done from
Visual Studio Team Explorer tab. Use *Clone Git Repository* and
provide address of your SaberMod repo. *Build All* action compiles
.dll modules. Other targets can be compiled from *CMake Targets View*
in *Project Explorer*.

CMake
-----

Install CMake for your operating system (Linux, Windows, MacOS and
others) and use ``CMakelists.txt`` script from the main SaberMod
directory.

Batch Scripts
-------------

Copy lcc.exe and q3asm.exe tools from *JK2 Editing Tools 2.0* into
bin/ directory. Run ``code/buildvms.bat`` batch script to generate
.qvm files.

License
=======

LCC 4.1 is Copyright (c) 1991-1998 by AT&T, Christopher W. Fraser and
David R. Hanson, and available under a non-copyleft license. You can
find it in ``code/tools/lcc/COPYRIGHT``. LCC version bundled with this
SDK comes from ioquake3 and it has been slightly modified by its
developers.

Some files in ``assets`` directory are modified assets from the
original, non-free JK2 1.04 release and licensed under *JK2 Editing
Tools 2.0* EULA.

Remaining parts of JK2 SDK GPL are licensed under GPLv2 as free
software. Read LICENSE.txt and README-raven.txt to learn
more. According to the license, among other things, you are obliged to
distribute full source code of your mod alongside of it, or at least a
written offer to ship it (eg a HTTP download link inside a .pk3
file). Moreover, any mod using patches from this repository **must**
be released under GPLv2 or a compatible license.

Q3ASM is Copyright (c) id Software and ioquake3 developers.

Authors
-------

* id Software 1999-2000
* Raven Software 1999-2002
* SaberMod developers 2015-2021

  + fau <witold.pilat@gmail.com> 2015-2021
  + Dziablo 2015-2016
  + Alereon 2019
  + TriForceX 2020-2021

Thanks
------

* Miso - Sending patches, testing, promoting SaberMod by hosting
  servers and events.
* Daggolin (boy) - Technical discussion, sharing patches and his JK2
  modding expertise.
* Xycaleth - Creating League mod that was a great inspiration to
  SaberMod and sharing its source code.
* ouned - Engine and modding expertise.
* Bucky, God, Kameleon, michl, PowTech, Tr!force - Providing valuable
  programming input, review, ideas and patches.
* Developers of jk2mv, mvsdk, Jedi Academy, OpenJK, ioq3, jomme, JA++
  (japp), League Mod and other open source id tech 3 mods for various
  code bugfixes.
* Players who help testing and improving SaberMod on a daily basis.

.. _GitHub : https://github.com/aufau/SaberMod
.. _`JK2 SDK GPL`: https://github.com/aufau/jk2sdk-gpl


CHANGELOG.rst
3.0
---

* Widescreen UI, HUD and FOV modes.
* New warmup system - wait for all players to ready up.
* Match referees promoted by admin or a popular vote.
* Match pause and timeouts.
* Unlagged CJK Instagib mode mimicking CJK mod mechanics.
* Last Man Standing gametype (11).
* Speed and movement crosshair indicators.
* Smooth camera enabled by default.
* Macro scan preventing kick script and other unfair binds.
* Anti-warp system detecting or preventing warping.
* More cheats blocked serverside.
* Spectator mode unlocking camera angles when following a player.
* Fast demo seeking with handy bindings.
* Lower CPU load on both server and client.
* Increased map and model limit in menus.
* Game modes organized in directories.
* Vampiric damage game modes.
* Automatically record demo and screenshot from a match.
* New opaque "dot" crosshair.
* A setting to make players pass through their teammates.
* A setting to allow pushing and pulling pickup items.
* A setting to make jump height independent of FPS.
* Fix for player connecting to a team gametype server causing lag.
* Fix for UI model selector going out of sync.
* Fix for the infamous player ghosting bug aka skin bug.
* Fix for flag not returning to base when falling to void on some maps.
* Fixes for multiple bugs when playing on a server with high uptime.
* Fixes many more issues.

2.0
---

* Unlagged Disruptor hit detection - lag compensation technique that
  allows playing instagib comfortably even with very high ping.
* New votes: Shuffle and Poll.
* Voting: callvote cooldown, ability to change vote.
* Callvote menu: Rules section, maps section: select from server maps
  rather than your own; download icon next to maps you're missing.
* 4 force kick effects to choose from: no effect, basejk, no damage,
  league mod.
* Crosshair: cg_crosshairColor, new "dot" crosshair (9).
* Ingame "Message of the Day", can be seen from ingame about menu.
* Fixed various issues with servers running for a few days or more.
* "players" game console command prints basic information about
  players and their fps and packets.
* Set original (97) FOV limit for all players with dmflags 16.
* Polished Clan Arena gametype and other previous features.
* Bug fixes and performance improvements - notably faster map restart.

1.0
---

* Clan Arena gametype.
* Votable game modes configured by server admin.
* Transparent duels - dueling players don't collide with others.
* Private duels - makes other players invisible (JK2MV 1.2+ servers).
* FPS-independent camera (cg_camerafps).
* Support for ent's "Jedi Knight Rewards 2.0".
* Voting: human-readable Gametype, Match Mode, Capturelimit.
* Improved logging: consistent format, timestamps, up to 4 logfiles
  with events customizable by bitmask cvars.
* Server commands: (un)lockteam, announce, say, tell, mode.
* cg_followKiller, cg_followPowerup, cg_drawTimer 2 (count down),
  cg_darkenDeadBodies, cg_drawClock, cg_duelGlow, cg_chatBeep,
  cg_privateDuel, cg_camerafps - most available in setup menu.
* g_instagib, g_spawnWeapons, g_spawnItems, g_infiniteAmmo,
  g_logFilter, g_consoleFilter, g_consoleFilter, g_modeDefault.
* League mod compatibility dmflag.
* Removed hi-res fonts as they are part of JK2MV now.
* Exploits blocked: speed hack, health reading, saber stealing,
  charged jump.
* Bugs fixed: glowing weapon, broken ydfa, rocket lock bug, saber lock
  bug, invisible saber, saber flying through the map on respawn,
  forcefield warping, 3-player duel and more.

Alpha
-----

* End-game statistics tuned for gametype.
* Detailed scoreboard tuned for gametype.
* Damage plums.
* Voting: No Kicks / With Kicks, Remove (removes to spectator),
  Teamsize, Roundlimit.
* Red Rover gametype.
* Smart follow button (+use).
* High resolution fonts.
* Smart name matching in game commands (tell, remove etc).
* /players client console command
* g_damagePlums, g_restrictChat, g_spawnShield, g_allowVote (uses
  bitmask now), teamsize, roundlimit, g_roundWarmup, g_teamsizeMin.
* Multiple bug fixes, optimizations and adjustments.


code/game/be_aas.h
/*
================================================================================
This file is part of SaberMod - Star Wars Jedi Knight II: Jedi Outcast mod.

Copyright (C) 1999-2000 Id Software, Inc.
Copyright (C) 1999-2002 Activision
Copyright (C) 2015-2021 Witold Pilat <witold.pilat@gmail.com>

This program is free software; you can redistribute it and/or modify it
under the terms and conditions of the GNU General Public License,
version 2, as published by the Free Software Foundation.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
================================================================================
*/

/*****************************************************************************
 * name:		be_aas.h
 *
 * desc:		Area Awareness System, stuff exported to the AI
 *
 * $Archive: /source/code/botlib/be_aas.h $
 * $Author: Mrelusive $
 * $Revision: 2 $
 * $Modtime: 10/05/99 3:32p $
 * $Date: 10/05/99 3:42p $
 *
 *****************************************************************************/

#ifndef MAX_STRINGFIELD
#define MAX_STRINGFIELD				80
#endif

//travel flags
#define TFL_INVALID				0x00000001	//traveling temporary not possible
#define TFL_WALK				0x00000002	//walking
#define TFL_CROUCH				0x00000004	//crouching
#define TFL_BARRIERJUMP			0x00000008	//jumping onto a barrier
#define TFL_JUMP				0x00000010	//jumping
#define TFL_LADDER				0x00000020	//climbing a ladder
#define TFL_WALKOFFLEDGE		0x00000080	//walking of a ledge
#define TFL_SWIM				0x00000100	//swimming
#define TFL_WATERJUMP			0x00000200	//jumping out of the water
#define TFL_TELEPORT			0x00000400	//teleporting
#define TFL_ELEVATOR			0x00000800	//elevator
#define TFL_ROCKETJUMP			0x00001000	//rocket jumping
#define TFL_BFGJUMP				0x00002000	//bfg jumping
#define TFL_GRAPPLEHOOK			0x00004000	//grappling hook
#define TFL_DOUBLEJUMP			0x00008000	//double jump
#define TFL_RAMPJUMP			0x00010000	//ramp jump
#define TFL_STRAFEJUMP			0x00020000	//strafe jump
#define TFL_JUMPPAD				0x00040000	//jump pad
#define TFL_AIR					0x00080000	//travel through air
#define TFL_WATER				0x00100000	//travel through water
#define TFL_SLIME				0x00200000	//travel through slime
#define TFL_LAVA				0x00400000	//travel through lava
#define TFL_DONOTENTER			0x00800000	//travel through donotenter area
#define TFL_FUNCBOB				0x01000000	//func bobbing
#define TFL_FLIGHT				0x02000000	//flight
#define TFL_BRIDGE				0x04000000	//move over a bridge
//
#define TFL_NOTTEAM1			0x08000000	//not team 1
#define TFL_NOTTEAM2			0x10000000	//not team 2

//default travel flags
#define TFL_DEFAULT	(TFL_WALK|TFL_CROUCH|TFL_BARRIERJUMP|TFL_JUMP|		\
		TFL_LADDER|TFL_WALKOFFLEDGE|TFL_SWIM|TFL_WATERJUMP|TFL_TELEPORT| \
		TFL_ELEVATOR|TFL_AIR|TFL_WATER|TFL_JUMPPAD|TFL_FUNCBOB)

typedef enum
{
	SOLID_NOT,			// no interaction with other objects
	SOLID_TRIGGER,		// only touch when inside, after moving
	SOLID_BBOX,			// touch on edge
	SOLID_BSP			// bsp clip, touch on edge
} solid_t;

//a trace is returned when a box is swept through the AAS world
typedef struct aas_trace_s
{
	qboolean	startsolid;	// if true, the initial point was in a solid area
	float		fraction;	// time completed, 1.0 = didn't hit anything
	vec3_t		endpos;		// final position
	int			ent;		// entity blocking the trace
	int			lastarea;	// last area the trace was in (zero if none)
	int			area;		// area blocking the trace (zero if none)
	int			planenum;	// number of the plane that was hit
} aas_trace_t;

/* Defined in botlib.h

//bsp_trace_t hit surface
typedef struct bsp_surface_s
{
	char name[16];
	int flags;
	int value;
} bsp_surface_t;

//a trace is returned when a box is swept through the BSP world
typedef struct bsp_trace_s
{
	qboolean		allsolid;	// if true, plane is not valid
	qboolean		startsolid;	// if true, the initial point was in a solid area
	float			fraction;	// time completed, 1.0 = didn't hit anything
	vec3_t			endpos;		// final position
	cplane_t		plane;		// surface normal at impact
	float			exp_dist;	// expanded plane distance
	int				sidenum;	// number of the brush side hit
	bsp_surface_t	surface;	// hit surface
	int				contents;	// contents on other side of surface hit
	int				ent;		// number of entity hit
} bsp_trace_t;
//
*/

//entity info
typedef struct aas_entityinfo_s
{
	int		valid;			// true if updated this frame
	int		type;			// entity type
	int		flags;			// entity flags
	float	ltime;			// local time
	float	update_time;	// time between last and current update
	int		number;			// number of the entity
	vec3_t	origin;			// origin of the entity
	vec3_t	angles;			// angles of the model
	vec3_t	old_origin;		// for lerping
	vec3_t	lastvisorigin;	// last visible origin
	vec3_t	mins;			// bounding box minimums
	vec3_t	maxs;			// bounding box maximums
	int		groundent;		// ground entity
	int		solid;			// solid type
	int		modelindex;		// model used
	int		modelindex2;	// weapons, CTF flags, etc
	int		frame;			// model frame number
	int		event;			// impulse events -- muzzle flashes, footsteps, etc
	int		eventParm;		// even parameter
	int		powerups;		// bit flags
	int		weapon;			// determines weapon and flash model, etc
	int		legsAnim;		// mask off ANIM_TOGGLEBIT
	int		torsoAnim;		// mask off ANIM_TOGGLEBIT
} aas_entityinfo_t;

// area info
typedef struct aas_areainfo_s
{
	int contents;
	int flags;
	int presencetype;
	int cluster;
	vec3_t mins;
	vec3_t maxs;
	vec3_t center;
} aas_areainfo_t;

// client movement prediction stop events, stop as soon as:
#define SE_NONE					0
#define SE_HITGROUND			1		// the ground is hit
#define SE_LEAVEGROUND			2		// there's no ground
#define SE_ENTERWATER			4		// water is entered
#define SE_ENTERSLIME			8		// slime is entered
#define SE_ENTERLAVA			16		// lava is entered
#define SE_HITGROUNDDAMAGE		32		// the ground is hit with damage
#define SE_GAP					64		// there's a gap
#define SE_TOUCHJUMPPAD			128		// touching a jump pad area
#define SE_TOUCHTELEPORTER		256		// touching teleporter
#define SE_ENTERAREA			512		// the given stoparea is entered
#define SE_HITGROUNDAREA		1024	// a ground face in the area is hit
#define SE_HITBOUNDINGBOX		2048	// hit the specified bounding box
#define SE_TOUCHCLUSTERPORTAL	4096	// touching a cluster portal

typedef struct aas_clientmove_s
{
	vec3_t endpos;			//position at the end of movement prediction
	int endarea;			//area at end of movement prediction
	vec3_t velocity;		//velocity at the end of movement prediction
	aas_trace_t trace;		//last trace
	int presencetype;		//presence type at end of movement prediction
	int stopevent;			//event that made the prediction stop
	int endcontents;		//contents at the end of movement prediction
	float time;				//time predicted ahead
	int frames;				//number of frames predicted ahead
} aas_clientmove_t;

// alternate route goals
#define ALTROUTEGOAL_ALL				1
#define ALTROUTEGOAL_CLUSTERPORTALS		2
#define ALTROUTEGOAL_VIEWPORTALS		4

typedef struct aas_altroutegoal_s
{
	vec3_t origin;
	int areanum;
	unsigned short starttraveltime;
	unsigned short goaltraveltime;
	unsigned short extratraveltime;
} aas_altroutegoal_t;

// route prediction stop events
#define RSE_NONE				0
#define RSE_NOROUTE				1	//no route to goal
#define RSE_USETRAVELTYPE		2	//stop as soon as on of the given travel types is used
#define RSE_ENTERCONTENTS		4	//stop when entering the given contents
#define RSE_ENTERAREA			8	//stop when entering the given area

typedef struct aas_predictroute_s
{
	vec3_t endpos;			//position at the end of movement prediction
	int endarea;			//area at end of movement prediction
	int stopevent;			//event that made the prediction stop
	int endcontents;		//contents at the end of movement prediction
	int endtravelflags;		//end travel flags
	int numareas;			//number of areas predicted ahead
	int time;				//time predicted ahead (in hundreth of a sec)
} aas_predictroute_t;
